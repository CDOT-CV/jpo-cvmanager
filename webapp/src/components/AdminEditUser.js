import React, { useEffect, useState } from "react";
import { Form } from "react-bootstrap";
import { useForm } from "react-hook-form";
import { Multiselect,DropdownList } from "react-widgets";
import EnvironmentVars from "../EnvironmentVars";

import "../components/css/Admin.css";
import "react-widgets/styles.css";

const AdminEditUser = (props) => {
  const [successMsg, setSuccessMsg] = useState("");
  const [selectedOrganizationNames, setSelectedOrganizationNames] = useState([]);
  const [selectedOrganizations, setSelectedOrganizations] = useState([]);
  const [organizationNames, setOrganizationNames] = useState([]);
  const [availableRoles, setAvailableRoles] = useState([]);
  const [apiData, setApiData] = useState({});
  const [errorState, setErrorState] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");
  const [submitAttempt, setSubmitAttempt] = useState(false);
  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
  } = useForm({
    defaultValues: {
      orig_email: "",
      email: "",
      first_name: "",
      last_name: "",
      super_user: "",
      organizations_to_add: [],
      organizations_to_modify: [],
      organizations_to_remove: [],
    },
  });

  const {
    userData,
    isLoginActive,
    setLoading,
    authLoginData
  } = props;

  const updateStates = (data) => {
    if (Object.keys(data).length !== 0) {
      let orgData = [];
      for (let i = 0; i < data.allowed_selections.organizations.length; i++) {
        let organization = data.allowed_selections.organizations[i];
        let temp = { id: i, name: organization };
        orgData.push(temp);
      }
      setOrganizationNames(orgData);

      let roleData = [];
      for (let i = 0; i < data.allowed_selections.roles.length; i++) {
        let role = data.allowed_selections.roles[i];
        let temp = { role: role };
        roleData.push(temp);
      }
      setAvailableRoles(roleData);

      setValue('orig_email', data.user_data.email);
      setValue('email', data.user_data.email);
      setValue('first_name', data.user_data.first_name);
      setValue('last_name', data.user_data.last_name);
      setValue('super_user', data.user_data.super_user);

      let tempOrganizations = [];
      let tempOrganizationNames = [];

      for (var i = 0; i < data.user_data.organizations.length; i++) {
        const org = data.user_data.organizations[i];
        let tempOrg = { id: i, name: org.name, role: org.role };
        let tempName = {id: i, name: data.user_data.organizations[i].name}
        tempOrganizations.push(tempOrg);
        tempOrganizationNames.push(tempName);
      }

      setSelectedOrganizations(tempOrganizations)
      setSelectedOrganizationNames(tempOrganizationNames)

    }
  };

  const fetchEditUserApiData = async (email) => {
    if (isLoginActive) {
      setLoading(true);
      setErrorState(false);
  
      try {
        const res = await fetch(EnvironmentVars.adminUser + '?user_email=' + email, {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: authLoginData["token"],
          },
        });
  
        const status = res.status;
        const data = await res.json();
  
        if (status === 200) {
          setApiData(data);
          updateStates(data);
        } else if (status === 500) {
          setErrorState(true);
        }
      } catch (exception_var) {
        setLoading(false);
        setErrorState(true);
        console.error(exception_var);
      }
    }
    setLoading(false);
  };

  useEffect(() => {
    fetchEditUserApiData(userData.email);
  }, [userData]);

  const updateOrganizations = (values) => {
    let newOrganizations = [];
    for (const name of values) {
      if (selectedOrganizations.some(e => e.name === name.name)){
        var index = selectedOrganizations.findIndex(function(item, i){
          return item.name === name.name
        });
        newOrganizations.push(selectedOrganizations[index]);
      } else if (!selectedOrganizations.some(e => e.name === name.name)) {
        name.role = availableRoles[0].role;
        newOrganizations.push(name);
      }
    }
  
    setSelectedOrganizations(newOrganizations)
    setSelectedOrganizationNames(values)
  }

  const sendPatchData = async (json) => {
    if (props.isLoginActive()) {
      props.setLoading(true);
      setErrorState(false);

      try {
        const res = await fetch(EnvironmentVars.adminUser, {
          method: "PATCH",
          headers: {
            "Content-Type": "application/json",
            Authorization: props.authLoginData["token"],
          },
          body: JSON.stringify(json),
        });
        const status = res.status;
        const data = await res.json();
        if (status === 200) {
          setSuccessMsg("Applied Changes Successfully!");
          props.updateUserData();
          resetMsg();
        } else if (status === 400) {
          setErrorMessage(data.message);
          setErrorState(true);
        } else if (status === 500) {
          setErrorMessage(data.message);
          setErrorState(true);
        }
      } catch (exception_var) {
        setErrorState(true);
        setErrorMessage(exception_var.message);
        console.error(exception_var);
      }
    }
    props.setLoading(false);
  };

  const organizationParser = (data, submitOrgs) => {
    let orgsToAdd = [];
    let orgsToModify = [];
    let orgsToRemove = [];

    for (const org of apiData.user_data.organizations){
      if (submitOrgs.some(e => e.name === org.name)){
        var index = submitOrgs.findIndex(function(item, i){
          return item.name === org.name;
        });
        if (submitOrgs[index].role !== org.role){
          const changedOrg = {name: submitOrgs[index].name, role: submitOrgs[index].role};
          orgsToModify.push(changedOrg);
        }
      } else {
        const removedOrg = {name: org.name, role: org.role};
        orgsToRemove.push(removedOrg);
      }
    }

    for (const org of submitOrgs){
      if (!apiData.user_data.organizations.some(e => e.name === org.name)){
        const newOrg = {name: org.name, role: org.role};
        orgsToAdd.push(newOrg);
      }
    }

    data.organizations_to_add = orgsToAdd;
    data.organizations_to_modify = orgsToModify;
    data.organizations_to_remove = orgsToRemove;
    return data;
  }


  const onSubmit = (data) => {
    if (selectedOrganizations.length !== 0) {
      setSubmitAttempt(false);
      let submitOrgs = selectedOrganizations;
      submitOrgs.forEach(elm => delete elm.id);
      data = organizationParser(data, submitOrgs);
      sendPatchData(data);
    } else {
      setSubmitAttempt(true);
    }
  };

  function resetMsg() {
    setTimeout(() => setSuccessMsg(""), 5000);
  }

  return (
    <div>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Group className="mb-3" controlId="email">
          <Form.Label>Email</Form.Label>
          <Form.Control
            type="email"
            placeholder="Enter user email"
            {...register("email", {
              required: "Please enter user email",
              pattern: {
                value: /^[^@ ]+@[^@ ]+\.[^@ .]{2,}$/,
                message: "Please enter a valid email",
              },
            })}
          />
          {errors.email && <p className="errorMsg">{errors.email.message}</p>}
        </Form.Group>

        <Form.Group className="mb-3" controlId="first_name">
          <Form.Label>First Name</Form.Label>
          <Form.Control
            type="text"
            placeholder="Enter user's first name"
            {...register("first_name", {
              required: "Please enter user's first name",
            })}
          />
          {errors.first_name && (
            <p className="errorMsg">{errors.first_name.message}</p>
          )}
        </Form.Group>

        <Form.Group className="mb-3" controlId="last_name">
          <Form.Label>Last Name</Form.Label>
          <Form.Control
            type="text"
            placeholder="Enter user's last name"
            {...register("last_name", {
              required: "Please enter user's last name",
            })}
          />
          {errors.last_name && (
            <p className="errorMsg">{errors.last_name.message}</p>
          )}
        </Form.Group>

        <Form.Group className="mb-3" controlId="super_user">
          <Form.Check
            label=" Super User"
            type="switch"
            {...register("super_user")}
          />
        </Form.Group>
        
        
        <Form.Group className="mb-3" controlId="organizations">
          <Form.Label>Organizations</Form.Label>
          <Multiselect
            className="form-multiselect"
            dataKey="name"
            textField="name"
            data={organizationNames}
            placeholder="Select organizations"
            value={selectedOrganizationNames}
            onChange={(value) => {
              updateOrganizations(value)
            }
            }
          />
        </Form.Group>

        {selectedOrganizations.length > 0 && (
          <Form.Group className="mb-3" controlId="roles">
            <Form.Label>Roles</Form.Label>
            <p className="spacer" />
            {selectedOrganizations.map((organization) => {              
              let role = {role: organization.role };

              return (
                <Form.Group className="mb-3" controlId={organization.id}>
                <Form.Label>{organization.name}</Form.Label>
                <DropdownList
                  className="form-dropdown"
                  dataKey="role"
                  textField="role"
                  data={availableRoles}
                  value={role}
                  onChange={(value) => {
                    role.role=value.role;
                    organization.role=value.role;
                  }}
                />
              </Form.Group>
              );
            })}
          </Form.Group>
        )}

        {selectedOrganizations.length === 0 && submitAttempt && (
          <p className="error-msg">Must select at least one organization</p>
        )}

        {successMsg && <p className="success-msg">{successMsg}</p>}
        {errorState && (
          <p className="error-msg">
            Failed to apply changes due to error: {errorMessage}
          </p>
        )}
        <div className="form-control">
          <label></label>
          <button type="submit" className="admin-button">Apply Changes</button>
        </div>
      </Form>
    </div>
  );
};

export default AdminEditUser;
