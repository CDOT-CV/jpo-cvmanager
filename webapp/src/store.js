import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
