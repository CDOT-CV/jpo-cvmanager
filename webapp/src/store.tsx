import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
