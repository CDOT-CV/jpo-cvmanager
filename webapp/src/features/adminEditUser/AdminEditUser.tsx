import React, { useEffect } from 'react'
import { Form } from 'react-bootstrap'
import { useForm } from 'react-hook-form'
import { Multiselect, DropdownList } from 'react-widgets'
import {
  selectSuccessMsg,
  selectSelectedOrganizationNames,
  selectSelectedOrganizations,
  selectOrganizationNames,
  selectAvailableRoles,
  selectErrorState,
  selectErrorMsg,
  selectSubmitAttempt,
  selectApiData,
  setSelectedRole,

  // actions
  getUserData,
  submitForm,
  updateOrganizations,
  UserApiDataOrgs,
} from './adminEditUserSlice'
import { useSelector, useDispatch } from 'react-redux'

import '../adminRsuTab/Admin.css'
import 'react-widgets/styles.css'
import { ThunkDispatch, AnyAction } from '@reduxjs/toolkit'
import { RootState } from '../../store'
import { Link, useParams } from 'react-router-dom'
import { getAvailableUsers, selectTableData } from '../adminUserTab/adminUserTabSlice'
import { ThemeProvider, Typography } from '@mui/material'
import { theme } from '../../styles'

const AdminEditUser = () => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const successMsg = useSelector(selectSuccessMsg)
  const selectedOrganizationNames = useSelector(selectSelectedOrganizationNames)
  const selectedOrganizations = useSelector(selectSelectedOrganizations)
  const organizationNames = useSelector(selectOrganizationNames)
  const availableRoles = useSelector(selectAvailableRoles)
  const apiData = useSelector(selectApiData)
  const errorState = useSelector(selectErrorState)
  const errorMsg = useSelector(selectErrorMsg)
  const submitAttempt = useSelector(selectSubmitAttempt)
  const userTableData = useSelector(selectTableData)
  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
  } = useForm({
    defaultValues: {
      orig_email: '',
      email: '',
      first_name: '',
      last_name: '',
      super_user: '',
      receive_error_emails: '',
      organizations_to_add: [],
      organizations_to_modify: [],
      organizations_to_remove: [],
    },
  })

  const { email } = useParams<{ email: string }>()

  useEffect(() => {
    if (
      (userTableData ?? []).find((user: AdminUserWithId) => user.email === email) &&
      Object.keys(apiData ?? {}).length == 0
    ) {
      console.log('getUserData')
      dispatch(getUserData(email))
    }
    console.log(
      'useEffect getUserData',
      email,
      userTableData,
      apiData,
      (userTableData ?? []).find((user: AdminUserWithId) => user.email === email),
      Object.keys(apiData ?? {}),
      Object.keys(apiData ?? {}).length,
      (userTableData ?? []).find((user: AdminUserWithId) => user.email === email) &&
        Object.keys(apiData ?? {}).length == 0
    )
  }, [email, userTableData, dispatch])

  useEffect(() => {
    dispatch(getAvailableUsers())
  }, [dispatch])

  useEffect(() => {
    if (apiData && Object.keys(apiData).length !== 0) {
      setValue('orig_email', apiData.user_data.email)
      setValue('email', apiData.user_data.email)
      setValue('first_name', apiData.user_data.first_name)
      setValue('last_name', apiData.user_data.last_name)
      setValue('super_user', apiData.user_data.super_user.toString())
      setValue('receive_error_emails', apiData.user_data.receive_error_emails.toString())
    }
    console.log('useEffect apiData', email, userTableData, apiData)
  }, [apiData, setValue])

  const onSubmit = (data: UserApiDataOrgs) => {
    dispatch(submitForm({ data }))
  }

  console.log('render', email, userTableData, apiData, Object.keys(apiData ?? {}).length)

  return (
    <div>
      {Object.keys(apiData ?? {}).length != 0 ? (
        <Form onSubmit={handleSubmit(onSubmit)}>
          <Form.Group className="mb-3" controlId="email">
            <Form.Label>Email</Form.Label>
            <Form.Control
              type="email"
              placeholder="Enter user email"
              {...register('email', {
                required: 'Please enter user email',
                pattern: {
                  value: /^[^@ ]+@[^@ ]+\.[^@ .]{2,}$/,
                  message: 'Please enter a valid email',
                },
              })}
            />
            {errors.email && (
              <p className="errorMsg" role="alert">
                {errors.email.message}
              </p>
            )}
          </Form.Group>

          <Form.Group className="mb-3" controlId="first_name">
            <Form.Label>First Name</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter user's first name"
              {...register('first_name', {
                required: "Please enter user's first name",
              })}
            />
            {errors.first_name && (
              <p className="errorMsg" role="alert">
                {errors.first_name.message}
              </p>
            )}
          </Form.Group>

          <Form.Group className="mb-3" controlId="last_name">
            <Form.Label>Last Name</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter user's last name"
              {...register('last_name', {
                required: "Please enter user's last name",
              })}
            />
            {errors.last_name && (
              <p className="errorMsg" role="alert">
                {errors.last_name.message}
              </p>
            )}
          </Form.Group>

          <Form.Group className="mb-3" controlId="super_user">
            <Form.Check label=" Super User" type="switch" {...register('super_user')} />
          </Form.Group>

          <Form.Group className="mb-3" controlId="receive_error_emails">
            <Form.Check label=" Receive Error Emails" type="switch" {...register('receive_error_emails')} />
          </Form.Group>

          <Form.Group className="mb-3" controlId="organizations">
            <Form.Label>Organizations</Form.Label>
            <Multiselect
              className="form-multiselect"
              dataKey="name"
              textField="name"
              data={organizationNames}
              placeholder="Select organizations"
              value={selectedOrganizationNames}
              onChange={(value) => {
                dispatch(updateOrganizations(value))
              }}
            />
          </Form.Group>

          {selectedOrganizations.length > 0 && (
            <Form.Group className="mb-3" controlId="roles">
              <Form.Label>Roles</Form.Label>
              <p className="spacer" />
              {selectedOrganizations.map((organization) => {
                let role = { role: organization.role }

                return (
                  <Form.Group className="mb-3" controlId={organization.id.toString()}>
                    <Form.Label>{organization.name}</Form.Label>
                    <DropdownList
                      className="form-dropdown"
                      dataKey="role"
                      textField="role"
                      data={availableRoles}
                      value={role}
                      onChange={(value) => {
                        dispatch(setSelectedRole({ ...organization, role: value.role }))
                      }}
                    />
                  </Form.Group>
                )
              })}
            </Form.Group>
          )}

          {selectedOrganizations.length === 0 && submitAttempt && (
            <p className="error-msg">Must select at least one organization</p>
          )}

          {successMsg && (
            <p className="success-msg" role="status">
              {successMsg}
            </p>
          )}
          {errorState && (
            <p className="error-msg" role="alert">
              Failed to apply changes due to error: {errorMsg}
            </p>
          )}
          <div className="form-control">
            <label></label>
            <button type="submit" className="admin-button">
              Apply Changes
            </button>
          </div>
        </Form>
      ) : (
        <Typography variant={'h4'} style={{ color: '#fff' }}>
          Unknown email address. Either this user does not exist, or you do not have permissions to view them.{' '}
          <Link to="../">Users</Link>
        </Typography>
      )}
    </div>
  )
}

export default AdminEditUser
