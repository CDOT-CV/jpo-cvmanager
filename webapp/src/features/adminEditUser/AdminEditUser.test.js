import React from "react";
import { render } from "@testing-library/react";
import AdminEditUser from "./AdminEditUser";
import { Provider } from "react-redux";
import { setupStore } from "../../store";

it("should take a snapshot", () => {
  const { asFragment } = render(
    <Provider store={setupStore({})}>
      <AdminEditUser userData={{}} />
    </Provider>
  );

  expect(
    asFragment(
      <Provider store={setupStore({})}>
        <AdminEditUser userData={{}} />
      </Provider>
    )
  ).toMatchSnapshot();
});
