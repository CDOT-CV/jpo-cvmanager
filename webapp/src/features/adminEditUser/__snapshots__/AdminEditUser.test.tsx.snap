// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-mocked-MuiTypography-root"
      style="color: rgb(255, 255, 255);"
    >
      Unknown email address. Either this user does not exist, or you do not have permissions to view them.
       
      <a
        href="/"
      >
        Users
      </a>
    </h4>
  </div>
</div>
`;
